vowpalwabbit/baseline_challenger_cb.cc

```diff
@@ -0,0 +1,196 @@
+        const CB::cb_class logged = (*lbl_example)->l.cb.costs[0];
+
+        double r = -logged.cost;
+        double w = (labelled_action == chosen_action ? 1 : 0) / logged.probability;
+        double w2 = (labelled_action == 0 ? 1 : 0) / logged.probability;
+
+        policy_expectation.update(w, r);
+        baseline.update(w2, r);
+      }
+
+      multiline_learn_or_predict<true>(base, examples, examples[0]->ft_offset);
+    }
+
+    //We play baseline if policy expectation is worse than the baseline lower bound
+    double ci = baseline.lower_bound();
+    double exp = policy_expectation.current();
+
+    //TODO don't check for it at every time step
+    bool play_baseline = ci > exp;
+
+    if(play_baseline) {
+      for(auto& it : action_scores) {
+        if (it.action == 0) {
+          it.action = chosen_action;
+        } else if (it.action == chosen_action) {
+          it.action = 0;
+        }
+      }
+    }
+  }
+
+};
+
+} // namespace VW
+
+template <bool is_learn>
+void learn_or_predict(baseline_challenger_data& data, multi_learner& base, multi_ex& examples)
+{
+  data.learn_or_predict<is_learn>(base, examples);
+}
+
+void save_load(baseline_challenger_data& data, io_buf& io, bool read, bool text)
+{
+  if (io.num_files() == 0) { return; }
+
+  data.policy_expectation.save_load(io, read, text, "policy");
+  data.baseline.save_load(io, read, text, "baseline");
+}
+
+void persist_metrics(baseline_challenger_data& data, metric_sink& metrics)
+{
+  if (!data.emit_metrics) {
+    return;
+  }
+  double ci = data.baseline.lower_bound();
+  double exp = data.policy_expectation.current();
+
+  metrics.float_metrics_list.emplace_back("baseline_cb_baseline_lowerbound", ci);
+  metrics.float_metrics_list.emplace_back("baseline_cb_policy_expectation", exp);
+  metrics.int_metrics_list.emplace_back("baseline_cb_baseline_in_use", ci > exp);
+}
+
+VW::LEARNER::base_learner* baseline_challenger_cb_setup(VW::setup_base_i& stack_builder)
+{
+  options_i& options = *stack_builder.get_options();
+  vw& all = *stack_builder.get_all_pointer();
+  double alpha;
+  double tau;
+  bool is_enabled = false;
+
+  option_group_definition new_options("Baseline challenger reduction: Build a CI around the baseline action and use it instead of the model if it's perfoming better");
+  new_options.add(make_option("baseline_challenger_cb", is_enabled).necessary().keep().help("Enable reduction"))
+      .add(make_option("cb_c_alpha", alpha).default_value(DEFAULT_ALPHA).keep().help("Confidence level for "))
+      .add(make_option("cb_c_tau", tau).default_value(DEFAULT_TAU).keep().help("Time constant for count decay"));
+
+  if (!options.add_parse_and_check_necessary(new_options)) {
+    return nullptr;
+  }
+
+  if (!options.was_supplied("cb_adf"))
+    { THROW("cb_challenger requires cb_explore_adf or cb_adf"); }
+
+  if (!all.logger.quiet)
+  {
+    *(all.trace_message) << "Using CB baseline challenger" << std::endl;
```

Comment:
This will already be conveyed in the `enabled reductions` component of the output so probably shouldnt be printed